core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AdaptiveScoreRingBuffer.java

```diff
@@ -0,0 +1,203 @@
+package ai.timefold.solver.core.impl.solver.termination;
+
+import java.util.Arrays;
+import java.util.Objects;
+
+import ai.timefold.solver.core.api.score.Score;
+
+import org.jspecify.annotations.NonNull;
+import org.jspecify.annotations.Nullable;
+
+final class AdaptiveScoreRingBuffer<Score_ extends Score<Score_>> {
```

Comment:
Is there a way to refactor the ring buffer to make it a general-purpose collection?
I can see it being useful to the Solver for more than this.

Reply:
Yes, but then we need to box the longs (unless you want to keep it as a map with long keys?) . Additionally it not a typical ring buffer (which normally acts a queue); its method removes all elements up to the floored query (and keeps the query entry). It also relies on the fact the queries will always increase due to the behaviour of `System.nanoTime`.

If you can spec out what you want the RingBuffer to be able to do, I can refactor it to match.

Reply:
Nevermind then - the generalization would hurt the primary purpose.
But the description of differences from a "typical" ring buffer would make for a nice Javadoc.